src/Analysis/Ast/Impl/Indexing/SymbolIndex.cs

```diff
@@ -0,0 +1,57 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using Microsoft.Python.Core;
+using Microsoft.Python.Parsing.Ast;
+
+namespace Microsoft.Python.Analysis.Indexing {
+    internal sealed class SymbolIndex : ISymbolIndex {
+        private readonly ConcurrentDictionary<Uri, IReadOnlyList<HierarchicalSymbol>> _index = new ConcurrentDictionary<Uri, IReadOnlyList<HierarchicalSymbol>>();
+        private bool _empty;
+        
+        public SymbolIndex() {
+            _empty = false;
+        }
+        
+        public bool isNotEmpty() {
+            return _empty;
+        }
+
+
+        public IEnumerable<HierarchicalSymbol> HierarchicalDocumentSymbols(Uri uri)
```

Comment:
Since you actually store lists, you can safely declare method as returning `IReadOnlyList`

Reply:
Again as said in my original PR with this code, the API surface will need to be rewritten to allow for the limiting to be done inside the index instead of outside. There would need to be some limit options passed down, which will need defining.